addons/account/models/account_payment.py

```diff
@@ -697,16 +719,46 @@ class payment_register(models.TransientModel):
     _description = 'Register Payment'
 
     payment_date = fields.Date(required=True, default=fields.Date.context_today)
-    journal_id = fields.Many2one('account.journal', required=True, domain=[('type', 'in', ('bank', 'cash'))])
-    payment_method_id = fields.Many2one('account.payment.method', string='Payment Method Type', required=True,
-                                        help="Manual: Get paid by cash, check or any other method outside of Odoo.\n"
-                                        "Electronic: Get paid automatically through a payment acquirer by requesting a transaction on a card saved by the customer when buying or subscribing online (payment token).\n"
-                                        "Check: Pay bill by check and print it from Odoo.\n"
-                                        "Batch Deposit: Encase several customer checks at once by generating a batch deposit to submit to your bank. When encoding the bank statement in Odoo, you are suggested to reconcile the transaction with the batch deposit.To enable batch deposit, module account_batch_payment must be installed.\n"
-                                        "SEPA Credit Transfer: Pay bill from a SEPA Credit Transfer file you submit to your bank. To enable sepa credit transfer, module account_sepa must be installed ")
+    journal_id = fields.Many2one(
+        'account.journal', required=True,
+        # FIXME: company only only if journal_id already set?
+        domain="""[('type', 'in', ('bank', 'cash')), ('company_id', '=', invoice_company_id)]"""
+    )
+    payment_method_id = fields.Many2one(
+        'account.payment.method', string='Payment Method Type', required=True,
+        domain="""[
+            (bool(journal_id), '=', 1),
+            '|',
+                '&', '&',
+                    (invoice_is_inbound, '=', 1),
+                    ('payment_type', '=', 'inbound'),
+                    ('id', 'in', journal_inbound_payment_method_ids),
+                '&', '&',
+                    (invoice_is_inbound, '!=', 1),
+                    ('payment_type', '=', 'outbound'),
+                    ('id', 'in', journal_outbound_payment_method_ids),
+        ]""",
```

Comment:
Qs:
* _valid_payment_method instead?
* if no journal is selected should we allow all methods or none?

Reply:
(original allows everything until a payment is set which seems odd)